dashboard/src/actions/overviewActions.js

```diff
@@ -407,3 +407,137 @@ const clearEditableFields = (item) => {
   item[CONSTANTS.IS_DIRTY_NAME] = false;
   item[CONSTANTS.IS_DIRTY_SERVER_DELETE] = false;
 };
+export const setMetadataModal = (isOpen, run) => async (dispatch, getState) => {
+  dispatch({
+    type: TYPES.SET_METADATA_MODAL,
+    payload: isOpen,
+  });
+};
+/**
+ * Function to get keySummary and send to parse data
+ * @function
+ * @return {Function} - dispatch the action and update the state
+ */
+export const getKeySummary = () => async (dispatch, getState) => {
+  try {
+    const endpoints = getState().apiEndpoint.endpoints;
+    const response = await API.get(uriTemplate(endpoints, "datasets_list"), {
+      params: { keysummary: true },
+    });
+    if (response.status === 200) {
+      if (response.data.keys) {
+        dispatch({
+          type: TYPES.GET_KEY_SUMMARY,
+          payload: response.data,
+        });
+        dispatch(parseKeySummaryforTree());
+      }
+    }
+  } catch (error) {
+    dispatch(showToast(DANGER, ERROR_MSG));
+  }
+};
+
+export const parseKeySummaryforTree = () => (dispatch, getState) => {
+  const parsedData = [];
+  const keySummary = getState().overview.keySummary.keys;
+  const optionRoles = Object.keys(keySummary);
+  const checkedItems = getState().overview.checkedItems;
+
+  for (const item of optionRoles) {
+    const dataObj = { title: item, options: [] };
+    for (const key in keySummary[item]) {
+      if (!isNonEssentialKey(key)) {
+        const aggregateKey = `${item}${CONSTANTS.KEYS_JOIN_BY}${key}`;
+        const isChecked = findItemCheked(aggregateKey, checkedItems);
+        const obj = constructTreeObj(aggregateKey, isChecked);
+        if (!keySummary[item][key]) {
+          dataObj.options.push(obj);
+        } /* has children */ else {
+          obj["children"] = constructChildTreeObj(
+            aggregateKey,
+            keySummary[item][key],
+            checkedItems,
+            dispatch
+          );
+          dataObj.options.push(obj);
+        }
+      }
+    }
+    parsedData.push(dataObj);
+  }
+  dispatch({
+    type: TYPES.SET_TREE_DATA,
+    payload: parsedData,
+  });
+};
+
+const constructChildTreeObj = (
+  aggregateKey,
+  entity,
+  checkedItems,
+  dispatch
+) => {
+  const childObj = [];
+  for (const item in entity) {
+    if (!isNonEssentialKey(item)) {
+      const isParentChecked = findItemCheked(aggregateKey, checkedItems);
+      const newKey = `${aggregateKey}${CONSTANTS.KEYS_JOIN_BY}${item}`;
+      const isChecked = isParentChecked || findItemCheked(newKey, checkedItems);
+      if (isChecked && !checkedItems.includes(newKey)) {
+        dispatch(updateCheckedItems(newKey));
+      }
```

Comment:
I'm confused and concerned by how this might be working.  We're updating the Redux store's instance of the checked items list, but we're not updating a local copy of it.  And, the update is dispatched to a task, so it doesn't happen right away.  So, every time we look at `checkedItems` we're looking at the state from the beginning of the parse.

It's probably OK that we're not keeping it up to date, and it's probably fine to defer the updates, but it seems inefficient to queue a pile of updates when, instead, we could just create a local copy of the state, update it as we go, and then update the Redux store _once_ at the end of the parse.

(And, if we pass in a reference to the list of checked items, then updates here would be visible in the caller; and, we wouldn't need to pass in `dispatch()` because the caller can make the call.)

Reply:
I'm not sure where to maintain the checkItems list during parsing